File: 351004/Shkredova/src/main/java/com/example/task310/service/NewsService.java
package com.example.task310.service;

import com.example.task310.dto.NewsRequestTo;
import com.example.task310.dto.NewsResponseTo;
import com.example.task310.exception.NotFoundException;
import com.example.task310.exception.ValidationException;
import com.example.task310.mapper.NewsMapper;
import com.example.task310.model.Creator;
import com.example.task310.model.Mark;
import com.example.task310.model.News;
import com.example.task310.repository.CreatorRepository;
import com.example.task310.repository.MarkRepository;
import com.example.task310.repository.NewsRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional
public class NewsService {
    private final NewsRepository newsRepository;
    private final CreatorRepository creatorRepository;
    private final MarkRepository markRepository;
    private final MarkService markService;  // Добавлено для автоматического создания меток
    private final NewsMapper mapper;

    @Transactional
    public NewsResponseTo create(NewsRequestTo request) {
        validateCreate(request);

        Creator creator = creatorRepository.findById(request.getCreatorId())
                .orElseThrow(() -> new NotFoundException("Creator not found with id: " + request.getCreatorId()));

        News entity = mapper.toEntity(request);
        entity.setCreator(creator);

        // Сначала сохраняем новость
        News saved = newsRepository.save(entity);

        // АВТОМАТИЧЕСКИ СОЗДАЁМ МЕТКИ С ID CREATOR
        String creatorId = String.valueOf(creator.getId());
        String[] colors = {"red", "green", "blue"};

        for (String color : colors) {
            String markName = color + creatorId;

            // Создаём метку, если её нет
            Mark mark = markRepository.findByName(markName)
                    .orElseGet(() -> {
                        Mark newMark = new Mark();
                        newMark.setName(markName);
                        return markRepository.save(newMark);
                    });

            // Привязываем к новости
            if (!saved.getMarks().contains(mark)) {
                saved.getMarks().add(mark);
            }
        }

        // Сохраняем новость с метками
        saved = newsRepository.save(saved);

        return mapper.toResponse(saved);
    }
    public List<NewsResponseTo> findAll() {
        return newsRepository.findAll().stream()
                .map(mapper::toResponse)
                .collect(Collectors.toList());
    }

    public NewsResponseTo findById(Long id) {
        return newsRepository.findById(id)
                .map(mapper::toResponse)
                .orElseThrow(() -> new NotFoundException("News not found with id: " + id));
    }

    public NewsResponseTo update(Long id, NewsRequestTo request) {
        News existing = newsRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("News not found with id: " + id));

        validateUpdate(request);

        existing.setTitle(request.getTitle());
        existing.setContent(request.getContent());

        if (request.getCreatorId() != null) {
            Creator creator = creatorRepository.findById(request.getCreatorId())
                    .orElseThrow(() -> new NotFoundException("Creator not found with id: " + request.getCreatorId()));
            existing.setCreator(creator);
        }

        try {
            News updated = newsRepository.save(existing);
            return mapper.toResponse(updated);
        } catch (DataIntegrityViolationException e) {
            throw new DataIntegrityViolationException("duplicate key value violates unique constraint");
        }
    }

    public void delete(Long id) {
        if (!newsRepository.existsById(id)) {
            throw new NotFoundException("News not found with id: " + id);
        }
        newsRepository.deleteById(id);
    }

    @Transactional
    public NewsResponseTo addMarkToNews(Long newsId, String markName) {
        News news = newsRepository.findById(newsId)
                .orElseThrow(() -> new NotFoundException("News not found with id: " + newsId));

        // Автоматически создаёт метку, если её нет
        Mark mark = markService.getOrCreateMark(markName);

        if (!news.getMarks().contains(mark)) {
            news.getMarks().add(mark);
            news = newsRepository.save(news);
        }

        return mapper.toResponse(news);
    }

    @Transactional
    public void removeMarkFromNews(Long newsId, String markName) {
        News news = newsRepository.findById(newsId)
                .orElseThrow(() -> new NotFoundException("News not found with id: " + newsId));

        markRepository.findByName(markName).ifPresent(mark -> {
            news.getMarks().remove(mark);
            newsRepository.save(news);
        });
    }

    @Transactional
    public void removeAllMarksFromNews(Long newsId) {
        News news = newsRepository.findById(newsId)
                .orElseThrow(() -> new NotFoundException("News not found with id: " + newsId));

        news.getMarks().clear();
        newsRepository.save(news);
    }

    @Transactional
    public void deleteMarkCompletely(String markName) {
        markRepository.findByName(markName).ifPresent(mark -> {
            // Отвязываем от всех новостей
            for (News news : mark.getNews()) {
                news.getMarks().remove(mark);
                newsRepository.save(news);
            }
            // Удаляем саму метку
            markRepository.delete(mark);
            System.out.println("🗑️ Метка полностью удалена: " + markName);
        });
    }

    @Transactional
    public void deleteNewsAndOrphanMarks(Long newsId) {
        // Находим новость
        News news = newsRepository.findById(newsId)
                .orElseThrow(() -> new NotFoundException("News not found with id: " + newsId));

        // Сохраняем список меток перед удалением
        List<Mark> marksToCheck = new ArrayList<>(news.getMarks());

        // Удаляем новость (связи удалятся автоматически)
        newsRepository.delete(news);

        // Проверяем каждую метку - если она больше не привязана ни к одной новости, удаляем
        for (Mark mark : marksToCheck) {
            if (mark.getNews().isEmpty()) {
                markRepository.delete(mark);
                System.out.println("🗑️ Удалена метка-сирота: " + mark.getName());
            }
        }
    }

    private void validateCreate(NewsRequestTo request) {
        if (request.getCreatorId() == null) {
            throw new ValidationException("Creator ID is required");
        }
        if (request.getTitle() == null || request.getTitle().trim().isEmpty()) {
            throw new ValidationException("Title is required");
        }
        if (request.getTitle().length() < 2 || request.getTitle().length() > 64) {
            throw new ValidationException("Title must be between 2 and 64 characters");
        }
        if (request.getContent() == null || request.getContent().trim().isEmpty()) {
            throw new ValidationException("Content is required");
        }
        if (request.getContent().length() < 4 || request.getContent().length() > 2048) {
            throw new ValidationException("Content must be between 4 and 2048 characters");
        }
    }

    private void validateUpdate(NewsRequestTo request) {
        if (request.getTitle() != null && (request.getTitle().length() < 2 || request.getTitle().length() > 64)) {
            throw new ValidationException("Title must be between 2 and 64 characters");
        }
        if (request.getContent() != null && (request.getContent().length() < 4 || request.getContent().length() > 2048)) {
            throw new ValidationException("Content must be between 4 and 2048 characters");
        }
    }
}

File: 351004/Leshok/Lab6/services/NewsService.js
// services/NewsService.js
const newsRepo = require('../repositories/NewsRepository');
const creatorRepo = require('../repositories/CreatorRepository');
const News = require('../models/News');
const { toResponse, fromRequest } = require('../dto/newsDto');
const { ValidationError, NotFoundError } = require('../utils/errors');

class NewsService {
  async getAll() {
    return newsRepo.findAll().map(toResponse);
  }

  async getById(id) {
    const news = newsRepo.findById(id);
    if (!news) throw new NotFoundError('News не найдена');
    return toResponse(news);
  }

  async create(newsRequest, currentUserLogin = null) {
    // Проверка, что creator существует
    const { title, content, creatorId } = fromRequest(newsRequest);
    const creator = creatorRepo.findById(creatorId);
    if (!creator) throw new ValidationError('Creator с таким id не существует', 400);
    
    // Если передан currentUserLogin (для v2), проверяем, что пользователь владелец или ADMIN
    if (currentUserLogin && creator.login !== currentUserLogin) {
      const user = creatorRepo.findByLogin(currentUserLogin);
      if (user.role !== 'ADMIN') throw new ValidationError('Вы можете создавать новости только от своего имени', 403);
    }
    
    const news = new News(null, title, content, creatorId);
    const saved = newsRepo.create(news);
    return toResponse(saved);
  }

  async update(id, newsRequest, currentUserLogin = null) {
    const existing = newsRepo.findById(id);
    if (!existing) throw new NotFoundError('News не найдена');
    
    // Проверка прав: только ADMIN или владелец новости
    if (currentUserLogin) {
      const creator = creatorRepo.findById(existing.creatorId);
      if (creator.login !== currentUserLogin) {
        const user = creatorRepo.findByLogin(currentUserLogin);
        if (user.role !== 'ADMIN') throw new ValidationError('Нет прав на редактирование этой новости', 403);
      }
    }
    
    const updateData = {};
    if (newsRequest.title !== undefined) updateData.title = newsRequest.title;
    if (newsRequest.content !== undefined) updateData.content = newsRequest.content;
    if (newsRequest.creatorId !== undefined) {
      const newCreator = creatorRepo.findById(newsRequest.creatorId);
      if (!newCreator) throw new ValidationError('Creator не найден', 400);
      updateData.creatorId = newsRequest.creatorId;
    }
    updateData.modified = new Date().toISOString();
    
    const updated = newsRepo.update(id, updateData);
    return toResponse(updated);
  }

  async delete(id, currentUserLogin = null) {
    const existing = newsRepo.findById(id);
    if (!existing) throw new NotFoundError('News не найдена');
    if (currentUserLogin) {
      const creator = creatorRepo.findById(existing.creatorId);
      if (creator.login !== currentUserLogin) {
        const user = creatorRepo.findByLogin(currentUserLogin);
        if (user.role !== 'ADMIN') throw new ValidationError('Нет прав на удаление этой новости', 403);
      }
    }
    return newsRepo.delete(id);
  }
}

module.exports = new NewsService();